src/packages/BooleanQueryTokenizer/BooleanQueryTokenizer.ts

```diff
@@ -0,0 +1,178 @@
+        [
+            TokenType.booleanOperator,
+            new RegExp(
+                `^(?:${andNotOperator}|${aroundOperator}|${simpleOperator})(?=${separatorOrEnd})`
+            ),
+        ],
+        [TokenType.parentheses, new RegExp(`^${parentheses}`)],
+    ];
+
+    /**
+     * Tokenizer splits an input text into meaningful chunks and labels each chunk according to its
+     * lexical meaning, suchwise producing a list of tokens.
+     */
+    public tokenize(inputString: string): Token[] {
+        const tokens: Token[] = [];
+        let remainingInputString = inputString;
+        while (remainingInputString.length > 0) {
+            const token = this.tokenizeCurrentChunk(remainingInputString);
+            if (token) {
+                // If token has been found, reduce the `remainingQueryString`
+                // on the length of the matched lexeme and add the token to the result.
+                const { lexeme } = token;
+                remainingInputString = remainingInputString.substring(lexeme.length);
+                tokens.push(token);
+            } else {
+                // If no tokens has been matched, mark current character as unknown token.
+                // If the previous token is also unknown, append current character to
+                // the previous token.
+                const prevToken = tokens[tokens.length - 1];
+                if (prevToken && prevToken.type === TokenType.unknown) {
+                    prevToken.lexeme += remainingInputString.substring(0, 1);
+                } else {
+                    tokens.push({
+                        type: TokenType.unknown,
+                        lexeme: remainingInputString.substring(0, 1),
+                    });
+                }
+                remainingInputString = remainingInputString.substring(1);
+            }
+        }
+        return tokens;
+    }
+
+    public extractWordsAndPhrases(inputString: string): string[] {
+        return (
+            this.tokenize(inputString)
+                // Filter only words and phrases
+                .filter((token) => token.type === TokenType.word || token.type === TokenType.phrase)
+                .map((wordOrPhrase) => {
+                    // `phrase` tokens require some dirty preprocessing
+                    // to remove quotes and possible inner wildcards.
+                    if (wordOrPhrase.type === TokenType.phrase) {
+                        return wordOrPhrase.lexeme.replace(/('|"|\*)/g, '').trim();
+                    }
+                    return wordOrPhrase.lexeme;
+                })
+                // After `phrase` tokens preprocessing, some strings can become empty,
+                // so we should remove them.
+                .filter((str) => str)
+        );
+    }
+
+    /**
+     * `tokenizeCurrentChunk` tries to find the best possible match for the current
+     * part of the input text
+     */
+    // eslint-disable-next-line class-methods-use-this
```

Comment:
The goal is to have all public methods of BooleanQueryTokenizer as non-static methods for simplicity.

Otherwise, the interface of the module would look like this:
```ts
BooleanQueryTokenizer.tokenize() // <- static
BooleanQueryTokenizer#extractWordsAndPhrases() // <- not static
```

But indeed, it makes more sense to move this `// eslint-disable-next-line class-methods-use-this` to `BooleanQueryTokenizer.tokenize()` directly.

Reply:
addressed